### ISSUE TITLE
Security issue with param escaping?

### ISSUE THREAD

**Issue by thekiur** [mysqljs/mysql#731]
Hi, im running node-mysql latest on node-latest.
Somebody using the acunetix vulnerability scanner has triggered this error:
UNKNOWN COLUMN '$acunetix' IN WHERE CLAUSE.
The query: SELECT id, email FROM accounts WHERE username = ?

How is this possible? Its very dangerous to our application, please respond quickly.

**Comment by SystemParadox** [comment]
Are there any circumstances where this would lead to an injection attack?

As far as I can work out so far this appears to only ever result in syntax errors.

**Comment by dresende** [comment]
@SystemParadox: I don't think so. The report seems to be badly explained and seems to be related to constructing SQL based on user input without any check.

Good usage:

```js
db.query("SELECT * FROM users WHERE id = ?", [ +req.params.id ], next);
```

No harm on that, casting forces it to be a number. Even if it wasn't a number and the `+` was omitted, it's just fine (or else you would have problems when UPDATing columns with binary data - there's tests for that).

The problem here seems to be with something more like:

```js
// BAD! BAD!
db.query("SELECT * FROM " + req.params.table + " WHERE ....", next);
```

**Comment by skeggse** [comment]
@SystemParadox yeah I just took a look at the formatting and escaping code. I don't see any way that passing unvalidated data to be interpolated into the query could result in an injection vulnerability. That said, without validation you can easily get a syntax error.

**Comment by jpginc** [comment]
Hi guys,

I've been looking into this recently and think that it could be a real issue for less security aware programmers. For example consider the following login query:

```JavaScript
db.query("SELECT * FROM users WHERE id = ? and password = ?", [ userInput.id userInput.password ], next);
```
say the userInput object is as follows:

`{"id":"test","password":{"password":"1"}}`

The resulting query will be:
```MySQL
SELECT * FROM users WHERE id = 'test' and password = `password` = '1'
```

the following section will always evaluate to true:

```MySQL
password = `password` = '1'
```

less of an issue is that a malicious user could discover valid column names by checking the error messages returned by an application.

To avoid this issue programmers would need to validate that the value being passed into the query function is not an Object. I saw that your supposed to 'opt-in' to enable the object parsing but this doesn't appear to be implemented? should this be pushed down to the sqlstring library?
